Add transaction endpoints

Ullaakut opened the pull request:
## Goal of this PR

Fixes #176 

This PR is currently a draft PR because we are missing some data to provide complete responses to the `GetTransactionResult` endpoint.

I still open this PR before moving on to something else so that this work is not forgotten and that we deal with this whenever we have some time.

## Checklist

- [x] Is on the right branch
- [x] Documentation is up-to-date
- [x] Tests are up-to-date

awfm9 commented on api/access/server.go:
```diff
@@ -160,12 +160,97 @@ func (s *Server) SendTransaction(ctx context.Context, in *access.SendTransaction
+	}
+
+	// We also need the height of the transaction we're looking at.
+	height, err := s.index.HeightForBlock(tx.ReferenceBlockID)
```
This is incorrect. As I wrote on Slack, the `ReferenceBlockID` has nothing to do with the block that includes the transaction. It's merely used for expiry of the transaction. As far as I know, there is no way to get the block for a transaction in such a way, so we will have to make one.

awfm9 commented on api/access/server.go:
```diff
@@ -160,12 +160,97 @@ func (s *Server) SendTransaction(ctx context.Context, in *access.SendTransaction
+		return nil, fmt.Errorf("could not retrieve block height: %w", err)
+	}
+
+	// Finally, if the transaction occurred after the current sealed height, we can say it's executed,
+	// otherwise that it is sealed. Pending, finalized and expired transactions are not supported by our
+	// API since we do not have access to pending and expired ones, and that finalized ones are also executed
+	// on our side, which takes precedence according to the aforementioned implementation in Flow-Go.
```
I don't think this is correct. When we don't have a transaction result, but we have a transaction body, we can consider that the transaction is either pending or finalized. If we consider that we only deal with finalized blocks in the DPS (as we should imo), then a transaction is:

- finalized if there is no transaction result;
- executed if there is a transaction result, but the block that contained the transaction is below sealed;
- sealed if there is a transaction result and the block is above or same as last sealed.

awfm9 commented on api/access/server.go:
```diff
@@ -160,12 +160,97 @@ func (s *Server) SendTransaction(ctx context.Context, in *access.SendTransaction
+		return nil, fmt.Errorf("could not retrieve last height: %w", err)
+	}
+
+	sealIDs, err := s.index.SealsByHeight(lastHeight)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve block seals: %w", err)
+	}
+
+	// We then go through all sealed blocks at the last height and look for the latest sealed height.
+	var sealedHeight uint64
+	for _, sealID := range sealIDs {
+		seal, err := s.index.Seal(sealID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve seal: %w", err)
+		}
+
+		h, err := s.index.HeightForBlock(seal.BlockID)
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve block height: %w", err)
+		}
+
+		if h > sealedHeight {
+			sealedHeight = h
+		}
+	}
```
This seems unnecessarily complex, and it also doesn't work for a frequent edge case, where a block won't contain any seals. In the Flow Go code, we work around this, by indexing the last seal for each block when its finalized, even if that last seal was in a previous block. We probably need to do something similar.

Ullaakut replied:
The best we can do is move this complexity around though, unless I'm missing smth. When building the index, to index the last sealed height, we need to look at all the seals that we got, lookup the height for their blockID and pick the highest one. This means that we have to do read operations in the index writer as well which I'm not fond of :/ Is there a better solution I'm missing here?

Currently in Flow they do something kind of weird:

```go
func (e *Engine) checkLastSealed(finalizedID flow.Identifier) error {
	// TODO: better to query seals from protocol state,
	// switch to state.Final().LastSealed() when available
	seal, err := e.sealsDB.ByBlockID(finalizedID)
	if err != nil {
		return fmt.Errorf("could not get the last sealed for the finalized block: %w", err)
	}

	blockID := seal.BlockID
	sealedCommit := seal.FinalState

	mycommit, err := e.execState.StateCommitmentByBlockID(e.unit.Ctx(), blockID)
	if errors.Is(err, storage.ErrNotFound) {
		// have not executed the sealed block yet
		// in other words, this can't detect execution fork, if the execution is behind
		// the sealing
		return nil
	}

	if err != nil {
		return fmt.Errorf("could not get my state commitment OnFinalizedBlock, blockID: %v", blockID)
	}

	if mycommit != sealedCommit {
		sealed, err := e.state.AtBlockID(blockID).Head()
		if err != nil {
			return fmt.Errorf("could not get sealed block when checkLastSealed: %v, err: %w", blockID, err)
		}

		return fmt.Errorf("execution result is different from the sealed result, height: %v, block_id: %v, sealed_commit: %x, my_commit: %x",
			sealed.Height, blockID, sealedCommit, mycommit)
	}

	return nil
}
```

And they do mention that they should switch and directly use the protocol state but unfortunately we can't take such shortcuts.

In another place (in the consensus nodes), this is what they do:

```go
	seal, err := c.seals.ByBlockID(finalizedBlockID)
	if err != nil {
		c.log.Fatal().Err(err).Msgf("could not retrieve seal for finalized block %s", finalizedBlockID)
	}
	lastSealed, err := c.headers.ByBlockID(seal.BlockID)
	if err != nil {
		c.log.Fatal().Err(err).Msgf("could not retrieve last sealed block %s", seal.BlockID)
	}

	// it's important to use atomic operation to make sure that we have correct ordering
	atomic.StoreUint64(&c.atomicLastSealedHeight, lastSealed.Height)
```

I see a few options but none of them seem great to me:

1. Add yet another index for SealsByBlockID and assume that any block that is at least at a height of 3 below the last indexed height is finalized. We can then do something like `lastSealedHeight := HeightForBlockID(SealByBlockID(BlockIDForHeight(LastHeight - 3)))` to get what we need.
2. When we index seals, read the height for the blocks they point to and store the highest one in a `LastSealed` height index (fairly simple, but means we do reads in the writer)
3. Keep the current complexity in the API rather than in the indexes

What makes the most sense to you guys? Am I missing smth obvious here? @awfm9 @Maelkum

awfm9 commented on api/access/server.go:
```diff
@@ -160,12 +160,97 @@ func (s *Server) SendTransaction(ctx context.Context, in *access.SendTransaction
+		StatusCode:   statusCode,
+		ErrorMessage: result.ErrorMessage,
+		Events:       convert.EventsToMessages(events),
+		BlockId:      tx.ReferenceBlockID[:],
```
This is incorrect. Just to hammer the point home, the `ReferenceBlockID` is _not_ the ID of the block that contained the transaction, it is _only_ used for expiry.

awfm9 reviewed (changes requested):
There are a few problems with the code as it is now.

However, I'm also starting to worry, on a higher level, about how much bloat we add to the index database. Maybe it would be a smarter idea to index the execution state and the protocol state separately. A lot of the stuff we now store in our index are already there in the same way in the protocol state. We could maybe create a completely converted, more compact version of the protocol state, with its own low-level API, and then keep the DPS to retrieving registers and such.

Ullaakut commented:
@awfm9 I addressed your suggestions by adding two indexes: `HeightForTransaction` and (last) `Sealed` (height). LMK if that works for you. I also did the other mentioned changes for the transaction status etc.

I'm personally not fond of the bloating we currently have and especially not of having reads in the index writer (due to having to index the last sealed height while not knowing the height of the seals we index without making reads). [See my comment here](https://github.com/optakt/flow-dps/pull/283#discussion_r671015611).

awfm9 commented on api/dps/index_test.go:
```diff
@@ -667,14 +762,17 @@ func TestIndex_ListSealsForHeight(t *testing.T) {
 type apiMock struct {
 	GetFirstFunc                  func(ctx context.Context, in *GetFirstRequest, opts ...grpc.CallOption) (*GetFirstResponse, error)
 	GetLastFunc                   func(ctx context.Context, in *GetLastRequest, opts ...grpc.CallOption) (*GetLastResponse, error)
+	GetSealedFunc                   func(ctx context.Context, in *GetSealedRequest, opts ...grpc.CallOption) (*GetSealedResponse, error)
```
`go fmt`?

awfm9 reviewed (commented):
This looks good to me, but I think you made things hard for yourself on the transaction status more than necessary.

Currently, when we replay from the LedgerWAL, we retrieve the `commit` for every block. This corresponds to the state commitment after the block has been executed. This means that the status for every transaction of every block we index is at least `EXECUTED`.

Taking this a bit further, since we have a _single_ indexed block per height we index, the block is already finalized. Since we already indexed a commit for the block, there must have been a seal with the commit in the protocol state, so it's sealed as well.

In conclusion, we exclusively work on the sealed part of the blockchain, so every transaction should have the `SEALED` status by default. The `Last` function can already be considered the last sealed height.

There is no need, imo, to keep track of which blocks were explicitly sealed by payload seals, as this was already done to introduce the commits into the protocol state for each block.

I would suggest we keep the same behaviour for the live feeder, which should simplify a lot of things.

Ullaakut commented:
@awfm9 Aha, I think I misinterpreted what you meant by "When we don't have a transaction result, but we have a transaction body, we can consider that the transaction is either pending or finalized."

And for the last sealed height, does this mean that when we receive blocks with seals, they should always have seals up to their own level? Or is it just that we know our blocks are sealed but we might not have the seals for them yet?

Ullaakut commented:
@awfm9 Should be good now. I removed the Sealed index and used Last instead, and now the status can only be executed or sealed. LMK if I missed something or it's good to merge 👍

awfm9 commented:
> @awfm9 Aha, I think I misinterpreted what you meant by "When we don't have a transaction result, but we have a transaction body, we can consider that the transaction is either pending or finalized."
> 
> And for the last sealed height, does this mean that when we receive blocks with seals, they should always have seals up to their own level? Or is it just that we know our blocks are sealed but we might not have the seals for them yet?

You didn't really misunderstand anything, I just hadn't looked at the details yet. It seems to me that this is a crucial design choice we have to make, and I think only working on the sealed state is probably the best option. We can change this relatively easily later on, so this is fine for now.

awfm9 commented on api/access/server.go:
```diff
@@ -161,12 +162,73 @@ func (s *Server) SendTransaction(ctx context.Context, in *access.SendTransaction
-	return nil, errors.New("not implemented")
+func (s *Server) GetTransaction(_ context.Context, in *access.GetTransactionRequest) (*access.TransactionResponse, error) {
+	txID := flow.HashToID(in.Id)
+	tb, err := s.index.Transaction(txID)
```
Would still use `tx` there as well :D

awfm9 commented on api/access/server.go:
```diff
@@ -161,12 +162,73 @@ func (s *Server) SendTransaction(ctx context.Context, in *access.SendTransaction
+	}
+
+	// We also need the height of the transaction we're looking at.
+	height, err := s.index.HeightForTransaction(tx.ID())
```
Would be better to use `txID` here, rather than computing the hash again.

awfm9 commented on api/access/server.go:
```diff
@@ -161,12 +162,73 @@ func (s *Server) SendTransaction(ctx context.Context, in *access.SendTransaction
+	blockID := block.ID()
+
+	statusCode := uint32(0)
+	if result != nil && result.ErrorMessage == "" {
```
Saying it once more, but I don't think we should have any `nil` checks in our code base. If there is a `nil` result, we should just consider the transaction not to not exist at the moment. And even if you want to do that check, I would use an `errors.Is` sentinel error check, rather than returning a `nil` value from `Result()`.

awfm9 reviewed (approved):
Still wonder about whether the checks you do are necessary at all, or if we can _only_ use `SEALED` as status. Anyway, we can make changes later.
